pkg/didcomm/protocol/didexchange/service.go

```diff
@@ -38,29 +49,231 @@ type Service struct {
 	store             storage.Store
 }
 
+//state defines a didcomm service state
+type state struct {
+	MsgType string            `json:"@type"`
+	ID      string            `json:"@id"`
+	Thread  *decorator.Thread `json:"~thread"`
+	Current string
+}
+
+//key struct prepares the composite key for state persitance and lookup
+type key struct {
+	//protocol is defined in the msg type url for example "didexchange"
+	protocol string
+	//version is derived from msg type as well
+	version string
+}
+
 // New return didexchange service
 func New(store storage.Store, prov provider) *Service {
 	return &Service{outboundTransport: prov.OutboundTransport(), store: store}
 }
 
 // Handle didexchange msg
 func (s *Service) Handle(msg dispatcher.DIDCommMsg) error {
-	// TODO add Handle logic
+	//populate request type, protocol, version required for composite key
+	msgType := strings.Split(msg.Type, "/")
+	pk := &key{
+		protocol: msgType[1],
+		version:  msgType[2],
+	}
+	switch msg.Type {
+	case connectionInvite:
+		invitation := &Invitation{}
+		err := json.Unmarshal(msg.Payload, invitation)
+		if err != nil {
+			return err
+		}
+		state := &state{invitation.Type, invitation.ID, nil, ""}
+		return s.invitation(state, pk)
+	case connectionRequest:
+		request := &Request{}
+		err := json.Unmarshal(msg.Payload, request)
+		if err != nil {
+			return err
+		}
+		state := &state{request.Type, request.ID, request.Thread, ""}
+		return s.request(state, pk)
+	case connectionResponse:
+		response := &Response{}
+		err := json.Unmarshal(msg.Payload, response)
+		if err != nil {
+			return err
+		}
+		state := &state{response.Type, response.ID, response.Thread, ""}
+		return s.response(state, pk)
+	case connectionAck:
+		ack := &Ack{}
+		err := json.Unmarshal(msg.Payload, ack)
+		if err != nil {
+			return err
+		}
+		state := &state{ack.Type, ack.ID, ack.Thread, ""}
+		return s.ack(state, pk)
+
+	default:
+		return errors.New("Message type  not supported")
+	}
+
+}
+
+func (s *Service) checkAndValidateState(state *state, pk *key) (string, error) {
+
+	if state.Thread == nil && state.MsgType == connectionInvite {
+		state.Thread = &decorator.Thread{ID: state.ID}
+	}
+	if state.Thread == nil && (state.MsgType == connectionRequest || state.MsgType == connectionResponse || state.MsgType == connectionAck) {
+		return "", errors.Errorf("Thread id in the %w is empty, therefor state cannot be validated", state.MsgType)
+	}
+
+	key := []string{state.Thread.ID, pk.protocol, pk.version}
```

Comment:
We should be able to use `pk.String()` to get this composite key

Reply:
added